File: algorithms/depth-first-traversal.ts
import { Node } from '../data-structures/binary-tree';

function depthFirstTraversal(node: Node): void {
	console.log(node.val);

	if (node.left) {
		depthFirstTraversal(node.left);
	}

	if (node.right) {
		depthFirstTraversal(node.right);
	}
}

const root = new Node(
	1,
	new Node(2, new Node(4), new Node(5)),
	new Node(3, null, new Node(6))
);

depthFirstTraversal(root);

File: algorithms/depth_first_traversal.go
package algorithms

import (
	"fmt"

	datastructs "github.com/claudealdric/dsa/data-structures"
)

func DfsRecursive(node *datastructs.TreeNode[int], nums *[]int) {
	if node == nil {
		return
	}

	*nums = append(*nums, node.Val)

	if node.Left == nil && node.Right == nil {
		return
	}

	DfsRecursive(node.Left, nums)
	DfsRecursive(node.Right, nums)
}

func DfsIterative(node *datastructs.TreeNode[int]) {
	var stack datastructs.Stack[*datastructs.TreeNode[int]]
	if node != nil {
		stack.Push(node)
	}

	for len(stack) > 0 {
		curr, _ := stack.Pop()
		fmt.Println(curr.Val)

		if curr.Right != nil {
			stack.Push(curr.Right)
		}
		if curr.Left != nil {
			stack.Push(curr.Left)
		}
	}
}
